[BUG] cache_test failures when testing with 128-bit decimal

While testing #4823 on Spark 3.2, cache tests were failing when 128-bit decimals were added to `decimal_gens`.  Specifically two tests were failing:
```
[2022-02-18T19:25:33.459Z] FAILED ../../src/main/python/cache_test.py::test_cache_partial_load[{'spark.sql.inMemoryColumnarStorage.enableVectorizedReader': 'true'}-Decimal(20,2)][ALLOW_NON_GPU(CollectLimitExec)]
[2022-02-18T19:25:33.459Z] FAILED ../../src/main/python/cache_test.py::test_cache_cpu_gpu_mixed[{'spark.sql.inMemoryColumnarStorage.enableVectorizedReader': 'true'}-Decimal(20,2)]
```

`test_cache_partial_load` was failing with "not all the plan is on the GPU" while `test_cache_cpu_gpu_mixed` was failing with:
```
E                   Caused by: java.lang.ArithmeticException: Decimal precision 21 exceeds max precision 20
E                   	at org.apache.spark.sql.types.Decimal.set(Decimal.scala:122)
E                   	at org.apache.spark.sql.types.Decimal$.apply(Decimal.scala:571)
E                   	at org.apache.spark.sql.types.Decimal.apply(Decimal.scala)
E                   	at org.apache.spark.sql.execution.vectorized.WritableColumnVector.getDecimal(WritableColumnVector.java:378)
E                   	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
E                   	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
E                   	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:757)
E                   	at org.apache.spark.sql.execution.collect.UnsafeRowBatchUtils$.encodeUnsafeRows(UnsafeRowBatchUtils.scala:80)
E                   	at org.apache.spark.sql.execution.collect.Collector.$anonfun$processFunc$1(Collector.scala:178)
E                   	at org.apache.spark.scheduler.ResultTask.$anonfun$runTask$3(ResultTask.scala:75)
E                   	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
E                   	at org.apache.spark.scheduler.ResultTask.$anonfun$runTask$1(ResultTask.scala:75)
E                   	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:55)
E                   	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:150)
E                   	at org.apache.spark.scheduler.Task.$anonfun$run$1(Task.scala:119)
E                   	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
E                   	at org.apache.spark.scheduler.Task.run(Task.scala:91)
E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$13(Executor.scala:813)
E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1620)
E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:816)
E                   	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
E                   	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:672)
E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E                   	... 1 more
```

--- a/integration_tests/src/main/python/cache_test.py
+++ b/integration_tests/src/main/python/cache_test.py
@@ -24,8 +24,7 @@ from marks import incompat, allow_non_gpu, ignore_order
 enable_vectorized_confs = [{"spark.sql.inMemoryColumnarStorage.enableVectorizedReader": "true"},
                            {"spark.sql.inMemoryColumnarStorage.enableVectorizedReader": "false"}]
 
-# cache does not work with 128-bit decimals, see https://github.com/NVIDIA/spark-rapids/issues/4826
-_cache_decimal_gens = [decimal_gen_32bit, decimal_gen_64bit]
+_cache_decimal_gens = [decimal_gen_32bit, decimal_gen_64bit, decimal_gen_128bit]
 _cache_single_array_gens_no_null = [ArrayGen(gen) for gen in all_basic_gens_no_null + _cache_decimal_gens]
 
 decimal_struct_gen= StructGen([['child0', sub_gen] for ind, sub_gen in enumerate(_cache_decimal_gens)])

--- a/sql-plugin/src/main/311+-all/scala/com/nvidia/spark/rapids/shims/v2/ParquetCachedBatchSerializer.scala
+++ b/sql-plugin/src/main/311+-all/scala/com/nvidia/spark/rapids/shims/v2/ParquetCachedBatchSerializer.scala
@@ -907,8 +907,10 @@ protected class ParquetCachedBatchSerializer extends GpuCachedBatchSerializer wi
           index -> inMemReqSparkSchema.fields.indexOf(reqSparkSchemaInCacheOrder.fields(index))
         }.toMap
 
-        val reqParquetSchemaInCacheOrder =
-          sparkToParquetSchemaConverter.convert(reqSparkSchemaInCacheOrder)
+        val reqParquetSchemaInCacheOrder = new org.apache.parquet.schema.MessageType(
+          inMemCacheParquetSchema.getName(), reqSparkSchemaInCacheOrder.fields.map { f =>
+            inMemCacheParquetSchema.getFields().get(inMemCacheParquetSchema.getFieldIndex(f.name))
+          }:_*)
 
         // reset spark schema calculated from parquet schema
         hadoopConf.set(ParquetReadSupport.SPARK_ROW_REQUESTED_SCHEMA, inMemReqSparkSchema.json)

--- a/sql-plugin/src/main/311+-all/scala/com/nvidia/spark/rapids/shims/v2/ParquetCachedBatchSerializer.scala
+++ b/sql-plugin/src/main/311+-all/scala/com/nvidia/spark/rapids/shims/v2/ParquetCachedBatchSerializer.scala
@@ -889,13 +889,9 @@ protected class ParquetCachedBatchSerializer extends GpuCachedBatchSerializer wi
         val inMemCacheSparkSchema = parquetToSparkSchemaConverter.convert(inMemCacheParquetSchema)
 
         val totalRowCount = parquetFileReader.getRowGroups.asScala.map(_.getRowCount).sum
-        val sparkToParquetSchemaConverter = new SparkToParquetSchemaConverter(hadoopConf)
         val inMemReqSparkSchema = StructType(selectedAttributes.toStructType.map { field =>
           inMemCacheSparkSchema.fields(inMemCacheSparkSchema.fieldIndex(field.name))
         })
-        val inMemReqParquetSchema = sparkToParquetSchemaConverter.convert(inMemReqSparkSchema)
-        val columnsRequested: util.List[ColumnDescriptor] = inMemReqParquetSchema.getColumns
-
         val reqSparkSchemaInCacheOrder = StructType(inMemCacheSparkSchema.filter(f =>
           inMemReqSparkSchema.fields.exists(f0 => f0.name.equals(f.name))))
 
@@ -912,20 +908,21 @@ protected class ParquetCachedBatchSerializer extends GpuCachedBatchSerializer wi
             inMemCacheParquetSchema.getFields().get(inMemCacheParquetSchema.getFieldIndex(f.name))
           }:_*)
 
+        val columnsRequested: util.List[ColumnDescriptor] = reqParquetSchemaInCacheOrder.getColumns
         // reset spark schema calculated from parquet schema
         hadoopConf.set(ParquetReadSupport.SPARK_ROW_REQUESTED_SCHEMA, inMemReqSparkSchema.json)
         hadoopConf.set(ParquetWriteSupport.SPARK_ROW_SCHEMA, inMemReqSparkSchema.json)
 
         val columnsInCache: util.List[ColumnDescriptor] = reqParquetSchemaInCacheOrder.getColumns
         val typesInCache: util.List[Type] = reqParquetSchemaInCacheOrder.asGroupType.getFields
-        val missingColumns = new Array[Boolean](inMemReqParquetSchema.getFieldCount)
+        val missingColumns = new Array[Boolean](reqParquetSchemaInCacheOrder.getFieldCount)
 
         // initialize missingColumns to cover the case where requested column isn't present in the
         // cache, which should never happen but just in case it does
-        val paths: util.List[Array[String]] = inMemReqParquetSchema.getPaths
+        val paths: util.List[Array[String]] = reqParquetSchemaInCacheOrder.getPaths
 
-        for (i <- 0 until inMemReqParquetSchema.getFieldCount) {
-          val t = inMemReqParquetSchema.getFields.get(i)
+        for (i <- 0 until reqParquetSchemaInCacheOrder.getFieldCount) {
+          val t = reqParquetSchemaInCacheOrder.getFields.get(i)
           if (!t.isPrimitive || t.isRepetition(Type.Repetition.REPEATED)) {
             throw new UnsupportedOperationException("Complex types not supported.")
           }

--- a/sql-plugin/src/main/311+-all/scala/com/nvidia/spark/rapids/shims/v2/ParquetCachedBatchSerializer.scala
+++ b/sql-plugin/src/main/311+-all/scala/com/nvidia/spark/rapids/shims/v2/ParquetCachedBatchSerializer.scala
@@ -53,7 +53,7 @@ import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference,
 import org.apache.spark.sql.catalyst.expressions.codegen.GenerateUnsafeProjection
 import org.apache.spark.sql.catalyst.util.{ArrayBasedMapData, ArrayData, GenericArrayData, MapData}
 import org.apache.spark.sql.columnar.CachedBatch
-import org.apache.spark.sql.execution.datasources.parquet.{ParquetReadSupport, ParquetToSparkSchemaConverter, ParquetWriteSupport, SparkToParquetSchemaConverter, VectorizedColumnReader}
+import org.apache.spark.sql.execution.datasources.parquet.{ParquetReadSupport, ParquetToSparkSchemaConverter, ParquetWriteSupport, VectorizedColumnReader}
 import org.apache.spark.sql.execution.datasources.parquet.rapids.shims.v2.{ParquetRecordMaterializer, ShimVectorizedColumnReader}
 import org.apache.spark.sql.execution.vectorized.{OffHeapColumnVector, WritableColumnVector}
 import org.apache.spark.sql.internal.SQLConf

--- a/sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XShims.scala
+++ b/sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XShims.scala
@@ -430,7 +430,7 @@ abstract class Spark31XShims extends Spark301until320Shims with Logging {
         }),
       GpuOverrides.exec[InMemoryTableScanExec](
         "Implementation of InMemoryTableScanExec to use GPU accelerated Caching",
-        ExecChecks((TypeSig.commonCudfTypes + TypeSig.DECIMAL_64 + TypeSig.STRUCT
+        ExecChecks((TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.STRUCT
             + TypeSig.ARRAY + TypeSig.MAP).nested(), TypeSig.all),
         (scan, conf, p, r) => new InMemoryTableScanMeta(scan, conf, p, r)),
       GpuOverrides.exec[ArrowEvalPythonExec](

